lib/Interpreter/IncrementalCUDADeviceCompiler.h

```diff
@@ -157,49 +116,78 @@ namespace cling {
     ///\param [in] invocationOptions - The invocationOptions of the interpreter.
     ///\param [in] intprOptLevel - The optimization level of the interpreter.
     ///\param [in] debugInfo - The debugInfo of the CompilerInstance.
+    ///\param [in] hostTriple - The llvm triple of the host system
     void setCuArgs(const clang::LangOptions& langOpts,
                    const cling::InvocationOptions& invocationOptions,
                    const int intprOptLevel,
-                   const clang::codegenoptions::DebugInfoKind debugInfo);
-
-    ///\brief Save .cu file, if cuda device code compiler failed at translation.
-    ///
-    ///\returns The error code of the rename operation
-    std::error_code saveFaultyCUfile();
+                   const clang::codegenoptions::DebugInfoKind debugInfo,
+                   const llvm::Triple hostTriple);
 
   public:
     ///\brief Constructor for IncrementalCUDADeviceCompiler
     ///
-    ///\param [in] filePath - All files will generated in the folder of the
-    ///       filePath, except the fatbin file, if it have another path. Have
-    ///       to end with a separator. Can be empty.
+    ///\param [in] fileBuffer - buffer in which the generated code of the
+    ///       transaction is stored
     ///\param [in] optLevel - The optimization level of the interpreter
-    /// instance.
+    ///       instance.
     ///       The value will be copied, because a change of it is not allowed.
     ///\param [in] invocationOptions - Contains values for the arguments of
-    ///       clang and the NVIDIA tool fatbinary.
+    ///       the interpreter instance and the NVIDIA tool fatbinary.
     ///\param [in] CI - Will be used for m_CuArgs and the include path handling.
     IncrementalCUDADeviceCompiler(
-        const std::string& filePath, const int optLevel,
-        const cling::InvocationOptions& invocationOptions,
+        std::shared_ptr<llvm::SmallVectorImpl<char>> fatbinBuffer,
+        const int optLevel, const cling::InvocationOptions& invocationOptions,
         const clang::CompilerInstance& CI);
 
+    ///\brief Returns a reference to the PTX interpreter
+    ///
+    ///\return std::unique_ptr< cling::Interpreter >&
+    ///
+    std::unique_ptr<Interpreter>& getPTXInterpreter() { return m_PTX_interp; }
```

Comment:
Why return the `unique_ptr`-ness of the interpreter? This allows the caller to do `getPTXInterpreter().resetI()`, and basically explicitly so (i.e. as an expected interface usage) because the returned `std::unique_ptr<Interpreter>&` is non-const. Suggest:
```suggestion
    Interpreter &getPTXInterpreter() { return *m_PTX_interp; }
```
if `m_PTX_interp` is guaranteed `!= nullptr`, else
```suggestion
    Interpreter *getPTXInterpreter() { return m_PTX_interp.get(); }
```